libs/common-services/src/main/java/uk/gov/di/ipv/core/library/service/IpvSessionService.java

```diff
@@ -51,52 +52,54 @@ public IpvSessionService(ConfigService configService) {
-        IpvSessionItem ipvSessionItem =
-                dataStore.getItemByIndex(
-                        "authorizationCode", DigestUtils.sha256Hex(authorizationCode));
-        return Optional.ofNullable(ipvSessionItem);
-    }
-
-    public IpvSessionItem getIpvSessionByAccessToken(String accessToken)
-            throws UnknownAccessTokenException, GetAccessTokenException {
-        try {
-            return Retry.runTaskWithBackoff(
-                    sleeper,
-                    7,
-                    10,
+    public IpvSessionItem getIpvSession(String ipvSessionId, boolean withRetry)
+            throws GetIpvSessionException {
+        if (withRetry) {
+            return callRunTaskWithBackoff(
                     () -> {
-                        var item =
-                                dataStore.getItemByIndex(
-                                        "accessToken", DigestUtils.sha256Hex(accessToken));
+                        var item = dataStore.getItem(ipvSessionId);
                         if (item == null) {
                             throw new RetryableException(
-                                    new UnknownAccessTokenException(
-                                            "The supplied access token was not found in the database"));
+                                    new GetIpvSessionException(
+                                            "The session not found in the database for the supplied session Id"));
                         }
                         return item;
                     });
-        } catch (InterruptedException e) {
-            LOGGER.warn(
-                    LogHelper.buildLogMessage(
-                            "getIpvSessionByAccessToken() backoff and retry sleep was interrupted"));
-            Thread.currentThread().interrupt();
-            throw new GetAccessTokenException("Failed to get the supplied access token", e);
-        } catch (NonRetryableException e) {
-            LOGGER.warn(
-                    LogHelper.buildErrorMessage(
-                            "getIpvSessionByAccessToken() exception occurred retrying getItemByIndex",
-                            e));
-            if (e.getCause() instanceof UnknownAccessTokenException uatException) {
-                throw uatException;
-            }
-            throw new GetAccessTokenException("Failed to get the supplied access token", e);
+        } else {
+            return dataStore.getItem(ipvSessionId);
         }
     }
 
+    public Optional<IpvSessionItem> getIpvSessionByAuthorizationCode(String authorizationCode)
+            throws GetIpvSessionException {
+        return callRunTaskWithBackoff(
+                () -> {
+                    IpvSessionItem ipvSessionItem =
+                            dataStore.getItemByIndex(
+                                    "authorizationCode", DigestUtils.sha256Hex(authorizationCode));
+                    return Optional.ofNullable(ipvSessionItem);
```

Comment:
Will this ever retry? It looks like it'll just return `Optional.empty` if it can't find it the first time...

It probably needs to look a bit more like `getIpvSessionByAccessToken` and return a non-optional `IpvSessionItem`.